devito/tools/dtypes_lowering.py

```diff
@@ -279,3 +310,14 @@ def infer_dtype(dtypes):
     else:
         # E.g., mixed integer arithmetic
         return max(dtypes, key=lambda i: np.dtype(i).itemsize, default=None)
+
+
+def dtype_to_petsctype(dtype):
+    """Map numpy types to PETSc datatypes."""
+
+    return {
+        np.int32: 'PetscInt',
+        np.float32: 'PetscScalar',
+        np.int64: 'PetscInt',
+        np.float64: 'PetscScalar'
+    }[dtype]
```

Comment:
Petsc use the same type for single and double precision?

Reply:
I believe PetscInt just represents an integer. Whether it is 32-bit or 64-bit depends on how you configured PETSc.

Reply:
I meant the float64 vs float32

Reply:
They use the same type for single and double precision but its representation depends on how the code was configured. https://petsc.org/release/manualpages/Sys/PetscScalar/

Reply:
So you can't have both single and double precision in the same code?

Reply:
No, as far as I'm aware PETSc doesn't support mixed precision